iceoryx_hoofs/test/moduletests/test_posix_named_semaphore.cpp

```diff
@@ -0,0 +1,266 @@
+    }
+}
+
+TEST_F(NamedSemaphoreTest, OpenExistingSemaphoreWorksWithoutDestroyingItInTheDtor)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "01274f45-74c0-43c2-95b1-f80afc78f017");
+
+    constexpr uint32_t INITIAL_VALUE = 1872U;
+    ASSERT_FALSE(NamedSemaphoreBuilder()
+                     .name(sutName)
+                     .openMode(OpenMode::PURGE_AND_CREATE)
+                     .permissions(sutPermission)
+                     .initialValue(INITIAL_VALUE)
+                     .create(sut)
+                     .has_error());
+
+    {
+        iox::cxx::optional<NamedSemaphore> sut2;
+        ASSERT_FALSE(NamedSemaphoreBuilder().name(sutName).openMode(OpenMode::OPEN_EXISTING).create(sut2).has_error());
+        EXPECT_TRUE(isSemaphoreValueEqualTo(*sut2, INITIAL_VALUE));
+    }
+
+    // if the dtor of sut2 unlinks the semaphore we should be unable to open it again
+    {
+        iox::cxx::optional<NamedSemaphore> sut2;
+        ASSERT_FALSE(NamedSemaphoreBuilder().name(sutName).openMode(OpenMode::OPEN_EXISTING).create(sut2).has_error());
+        EXPECT_TRUE(isSemaphoreValueEqualTo(*sut2, INITIAL_VALUE));
+    }
+
+    // verify that the original semaphore which has the ownership still works
+    EXPECT_TRUE(isSemaphoreValueEqualTo(*sut, INITIAL_VALUE));
+}
+
+TEST_F(NamedSemaphoreTest, OpenNonExistingSemaphoreFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ec124aaf-aad1-4375-a288-b3b06dbc6ac2");
+
+    auto result =
+        NamedSemaphoreBuilder().name(sutName).openMode(OpenMode::OPEN_EXISTING).permissions(sutPermission).create(sut);
+
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(SemaphoreError::NO_SEMAPHORE_WITH_THAT_NAME_EXISTS));
+}
+
+TEST_F(NamedSemaphoreTest, ExclusiveCreateFailsWhenSemaphoreAlreadyExists)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ec2c4ab4-dc63-4a4c-b66f-fded785ef4f0");
+
+    ASSERT_FALSE(NamedSemaphoreBuilder()
+                     .name(sutName)
+                     .openMode(OpenMode::PURGE_AND_CREATE)
+                     .permissions(sutPermission)
+                     .create(sut)
+                     .has_error());
+
+    iox::cxx::optional<NamedSemaphore> sut2;
+    auto result = NamedSemaphoreBuilder()
+                      .name(sutName)
+                      .openMode(OpenMode::EXCLUSIVE_CREATE)
+                      .permissions(sutPermission)
+                      .create(sut2);
+
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(SemaphoreError::ALREADY_EXIST));
+}
+
+TEST_F(NamedSemaphoreTest, SemaphoreWithInvalidNameFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "07ef3fdf-0e22-45f8-bb6b-68647f1211b6");
+
+    auto result =
+        NamedSemaphoreBuilder().name("///").openMode(OpenMode::PURGE_AND_CREATE).permissions(sutPermission).create(sut);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(SemaphoreError::INVALID_NAME));
+}
+
```

Comment:
Does it makes sense to add tests where you vary the permissions? e.g. `create` with `OpenMode::PURGE_AND_CREATE` and `iox::cxx::perms::none` should fail with `PERMISSION_DENIED`

Reply:
The problem is the user always has the permission to delete its own files. What would work would be to create a semaphore with a different user and only owner permissions then `PURGE_AND_CREATE` should fail when it is used with a different user.

But this is something we are unable to do in a unit test.